### ISSUE TITLE
returning an error from a custom importer?

### ISSUE THREAD

**Issue by chriseppstein** [sass/node-sass#651]
Is it possible to return an error from a custom importer instead of just calling `done({})` and getting the standard import not found error?

**Comment by am11** [comment]
Something like this: https://github.com/sass/node-sass/issues/632? The issue was opened with libsass (https://github.com/sass/libsass/issues/862), but it was closed there as well.

**Comment by chriseppstein** [comment]
Yes, but then it took a left turn. Sometimes an importer can decide definitively that the import doesn't exist. In this case it should be able to return an error with details about why.

Something like `done({error: "<some file> is not readable"})`.

**Comment by am11** [comment]
I agree, but this would need to be supported by libsass. We do not override errors returned by libsass.

The current behavior is as follow:

```javascript
var options = {
    data: '@import "dev-null";',
    importer: function(url, prev, done)  {
      done({file: url + ' is unavailable.'});
    },
    error: function(err) {
      console.error(err);
    }
};

require('node-sass').render(options);
```

Prints:

```javascript
{ status: 1,
  file: 'stdin',
  line: 1,
  column: 9,
  message: 'file to import not found or unreadable: dev-null is unavailable.\nCurrent dir: ',
  code: 1 }
```

**Comment by chriseppstein** [comment]
So what do we need from libsass to make this work? @hcatlin

**Comment by am11** [comment]
AFAICT, that is the standard `missing import` error which LibSass returns. 
The fact is your `{error: '..'}` syntax is more verbose and brings clarity of intent, for which [`struct Sass_Import`](https://github.com/sass/libsass/wiki/API-Sass-Importer-Internal) would probably need to have a `char* error` to support custom error message which will override the default message.

@mgreter can probably suggest a better solution here.

**Comment by mgreter** [comment]
It's a missing feature (one of the reasons I marked it as an experimental feature)! Custom functions can return a Sass_Error which is just another Sass_Value type. Custom importers on the other hand are expected to return a specific struct (created by sass_make_import_list). Maybe it would make sense to add imports as another Sass_Value type to be able to also return a Sass_Error value. Currently you cannot trigger a specific error via custom importers. @chriseppstein feel free to take the lead here!

**Comment by am11** [comment]
> Maybe it would make sense to add imports as another Sass_Value type to be able to also return a Sass_Error value.

:+1:

**Comment by chriseppstein** [comment]
@am11 I don't think it makes sense to return a `Sass_Value` here. It would be nonsensical to return a string or number, etc. A string message and way to provide a host-language stack trace would be sufficient. `error_message` and `error_details`?

**Comment by am11** [comment]
@chriseppstein, I have not yet actually looked into any aspect of Sass_Value or custom functions. Once [v2.0.0 binary matter](https://github.com/sass/node-sass/issues/602) is sorted out and released, I will dive deeper.

The core problem is to display the exact error message which user intended. From the example above, instead of:

> 'file to import not found or unreadable: dev-null is unavailable.\nCurrent dir: ',

the error should be:

> dev-null is unavailable.',

Having said that, what I think we have at least three approaches here:

## Approach 1:

Consider this importer code:

```javascript
importer: function(url, prev, done)  {
  return [
    {file: process1(url, prev)},
    {file: process2(url, prev)}
  ];
},
```

This is returning an array. I just wanted to highlight that for a single import, we can have multiple items returned by the importer.

We can add `char* error_message` in `struct Sass_Import` and change the behavior so if the `error_message` is found for any returned `Sass_Import` object, return the error message instead of standard missing import and stop further parser/lexer execution.

Catch: This would require change in our binding code and LibSass as well.
Cons: Except for dissimilarity from custom function, nothing comes to mind.

## Approach 2:

What @mgreter suggested
We can add inherit `Sass_Import` from `Sass_Value` so our binding can detect which object to instantiate (error or value).

Catch: That would also require change in our binding and LibSass code.
Cons: I do not know how this is nonsense, I find it precise and a step towards convergence of this newly introduces Sass type-system. 

## Approach 3:

See https://github.com/sass/node-sass/issues/646. As [you can tell](https://github.com/sass/node-sass/issues/646#issuecomment-72423827) I am not 100% sure how it works, but what I understood is that:

* `@debug`, `@error`, `@warn` will have corresponding callbacks in our binding.
* we will maintain vectors with lifetime of compile.
* on_compile_result_ready, we will serialise the those vectors into our result object (JSON-like JS object literal).

Using that machinery, we can detect if `importer` has returned error, we can call binding from our API (JS) code to add string to error vector. And skip return `null` to LibSass for this import (with `error`). This way we will be able to return the intended error message.

Catch: It would require zero change in LibSass and we can fix it from our binding code. However, I am not sure if LibSass would consider it as a successfully compilation in presence of `@error` which is not what we want.

Cons: I think this goes against the [single responsibility principle](http://www.oodesign.com/single-responsibility-principle.html).

**Comment by chriseppstein** [comment]
I think approach 1 is the only one that makes sense to me.

Approach 2 is bad because the goal of importing is to return a string that can be evaluated as sass or scss code. As I understand it, a Sass_Value is a value that is returned from a script expression. While under the covers in libsass an import would result in a Sass error and would fire off that machinery, but it doesn't mean that the importer should be the one making that error. Alternatively, we could define a different kind of result where Sass Errors are an object that can be returned to script or to import but are not part of the Sass_Value data type and could then be shared in both cases.

**Comment by chriseppstein** [comment]
https://github.com/sass/libsass/issues/880 adds this capability to libsass can we expose this to node-sass?

**Comment by am11** [comment]
We only update submodules when libsass tagged version is out. The latest released version is v3.1.0, which we are using in node-sass v2.0.1.
Going by the situation here: https://github.com/sass/libsass/issues/812#issuecomment-77966156, I think we can make an alpha release with libsass v3.1.0 (to mitigate io.js versions compatibility complains). Once libsass 3.2 is out we will implement these new API features. Hopefully this will happen in node-sass v3.0.0-beta.

**Comment by am11** [comment]
This is addressed by bd51ea4 via #817.
In order to throw error from importer, use:

```javascript
importer: function(url, prev, done) {
  done(new Error('my preferred error message for missing files'));
}
// or
importer: function() {
  return new Error('my preferred error message for missing files');
}
```

This feature will be made available in next beta.

Thanks for your cooperation! (-8

**Comment by Snugug** [comment]
Yay! Thanks!
